### PR TITLE
Restful api

### PR DESCRIPTION
**Pull request by I-codebyte**

### CODE REVIEW

**Comment by I-codebyte** [comment]
Assignment on restful api

File: package.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "techy-assignment",
+  "version": "1.0.0",
+  "description": "this file is used to create a bank account class and transaction class.\r main functions are: 1. to create user, 2. to deposit, 3. to withdraw, 4. to view transaction history\r New features: withdrawlimit and transfer between to accounts",
+  "main": "bank.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "server": "node restapi.js"
+  },
```
Comment (joshuaunity) [inline package.json:9]: add a start command

File: restapi.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+mongoose.connect(DBUri)
+.then(() => {
+    console.log(`Database is connected`)
+})
+.catch(() => {
+    console.log(`Database is not connected`)
+});
+
+// Create an account
+app.post('/Createaccount/', async (req, res) => {
```
Comment (joshuaunity) [inline restapi.js]: avoid using CAPS in your API url's

Comment (joshuaunity) [reply restapi.js]: Also there are no validations in the APIs as there are a couple of them that needs a validation. Tis create account is one. You can also add comments like an example to show what the request body will be like for an API.

File: package.json (hunk 1 above)
Comment (I-codebyte) [reply package.json:9]: Done

File: restapi.js (hunk 2 above)
Comment (I-codebyte) [reply restapi.js]: Done

File: package.json (hunk 1 above)
Comment (joshuaunity) [reply package.json:9]: OK great 👍🏾

File: restapi.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+// Create an account
+/* e.g;
+{
+    "accountNumber": 3678299923,
+    "firstName": "Atinuke",
+    "lastName": "Gbolagade"
+}
+*/
+app.post('/createaccount/', async (req, res) => {
+    try {
```
Comment (joshuaunity) [inline restapi.js:37]: there is no validation here, yes the mongoose will return an error if I put numbers in the first and last name, but you should handle that error. also if I pass an empty string as the request there is also no validation.

File: restapi.js (hunk 2 above)
Comment (joshuaunity) [reply restapi.js]: Yes you have done some but there is still some left, I already made another comment on that area.

File: restapi.js (hunk 3 above)
Comment (I-codebyte) [reply restapi.js:37]: done

**Comment by joshuaunity** [comment]
Hi @IGNORAMUS1 I just saw you added the API collection, I have used docker to set up your server but there is an error when I try to create an account, please try on your end. Below is the error I get when I hit the createaccount API.
Screenshot 2024-05-22 172729

**Comment by joshuaunity** [comment]
Also, i made some modifications to the code as I couldn't connect to the DB. NOTE if your db is running in docker and your project is also running in docker you will need to make this modification to your URI

const DBUri = "mongodb://host.docker.internal:27017/Users"

Note the difference, instead of "localhost" I used "host.docker.internal" that's how you can make to docker applications communicate to each other.

I will also mention this in today's class.

**Comment by I-codebyte** [comment]
> Hi @IGNORAMUS1 I just saw you added the API collection, I have used docker to set up your server but there is an error when I try to create an account, please try on your end. Below is the error I get when I hit the createaccount API. Screenshot 2024-05-22 172729

I'm sorry for the late reply sir. I've not been able to follow up due to some issues.

I can't view the image u sent sir

**Comment by I-codebyte** [comment]
> Also, i made some modifications to the code as I couldn't connect to the DB. NOTE if your db is running in docker and your project is also running in docker you will need to make this modification to your URI
> 
> const DBUri = "mongodb://host.docker.internal:27017/Users"
> 
> Note the difference, instead of "localhost" I used "host.docker.internal" that's how you can make to docker applications communicate to each other.
> 
> I will also mention this in today's class.

Is this all I need to do for now sir?

**Comment by joshuaunity** [comment]
Hi @IGNORAMUS1, sorry for the late feedback. Here is a link to the image in case you ahre having issues viewing it on your end

https://private-user-images.githubusercontent.com/45713692/332879938-4e05511b-ef4c-480a-be59-0d35390dd112.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxOTg4ODcsIm5iZiI6MTcxNzE5ODU4NywicGF0aCI6Ii80NTcxMzY5Mi8zMzI4Nzk5MzgtNGUwNTUxMWItZWY0Yy00ODBhLWJlNTktMGQzNTM5MGRkMTEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDIzMzYyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOTY4MmQ0NmYzMGRhMmVlOTFhZGE0ODUxZGFkZDYxYjA0YjcyNDgzOTBjMWYwNjI0OTk0YTRhYTk3MzAzOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rO2SaiPo2Pf_ed_v9ruNB6vBvQxbCFRA0WuEnuNmHeY

**Comment by I-codebyte** [comment]
> Hi @IGNORAMUS1, sorry for the late feedback. Here is a link to the image in case you ahre having issues viewing it on your end
> 
> https://private-user-images.githubusercontent.com/45713692/332879938-4e05511b-ef4c-480a-be59-0d35390dd112.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxOTg4ODcsIm5iZiI6MTcxNzE5ODU4NywicGF0aCI6Ii80NTcxMzY5Mi8zMzI4Nzk5MzgtNGUwNTUxMWItZWY0Yy00ODBhLWJlNTktMGQzNTM5MGRkMTEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMxVDIzMzYyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOTY4MmQ0NmYzMGRhMmVlOTFhZGE0ODUxZGFkZDYxYjA0YjcyNDgzOTBjMWYwNjI0OTk0YTRhYTk3MzAzOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rO2SaiPo2Pf_ed_v9ruNB6vBvQxbCFRA0WuEnuNmHeY


The link is not loading at all. 


But this is how I connected my database and the restapi:
I put my yaml file for database in a separate folder which I named 'DBserver' so that when I do 'docker-compose up build' for the main server the yaml file for my database will be ignored.

So after I've run docker-compose up --build for the main server, I will then change my directory to '/DBserver' so that I can do 'docker-compose up --build' for my database yaml file too.


I will then on them separately on docker.

**Comment by joshuaunity** [comment]
Hi @IGNORAMUS1,

The initial error has been resolved. Here are my review highlights:

The system is not generating an account number.
The "Get Single Account" API is returning all accounts instead of just one.
When using the sample request in your Postman collection, I was unable to deposit into an account. Are you sure the correct request is specified for that API? I encountered the same issue with the withdrawal API.
There is a missing API for transferring money to another existing account.

Well done so far, you are almost done with this and viable to receive the certificate.

**Comment by I-codebyte** [comment]
> Hi @IGNORAMUS1,
> 
> The initial error has been resolved. Here are my review highlights:
> 
> The system is not generating an account number. The "Get Single Account" API is returning all accounts instead of just one. When using the sample request in your Postman collection, I was unable to deposit into an account. Are you sure the correct request is specified for that API? I encountered the same issue with the withdrawal API. There is a missing API for transferring money to another existing account.
> 
> Well done so far, you are almost done with this and viable to receive the certificate.

The Get Single Account, withdraw and deposit API are all working fine on my end sir. Maybe u are using Get All User API for Get Single Account API.

**Comment by joshuaunity** [comment]
Hi @IGNORAMUS1 i have reviewed the project, well done so far, all has been implemented but there are currently two issues

1. the "All transactions" API only returns an array of ID's rather than an array of the transactions themselves
2. I can create multiple accounts with the same first and last name, there should be validation against that.

**Comment by I-codebyte** [comment]
> Hi @IGNORAMUS1 i have reviewed the project, well done so far, all has been implemented but there are currently two issues
> 
> 1. the "All transactions" API only returns an array of ID's rather than an array of the transactions themselves
> 2. I can create multiple accounts with the same first and last name, there should be validation against that.

I don't want the all transactions to so long that's was why I made it to return only the IDs.
And also two user can have the same name that was why I put no restrictions on that too. If you want me to change it, I will do that once u give me the go ahead. 
Thanks sir

**Comment by joshuaunity** [comment]
> > Hi @IGNORAMUS1 i have reviewed the project, well done so far, all has been implemented but there are currently two issues
> > 
> > 1. the "All transactions" API only returns an array of ID's rather than an array of the transactions themselves
> > 2. I can create multiple accounts with the same first and last name, there should be validation against that.
> 
> I don't want the all transactions to so long that's was why I made it to return only the IDs.
> And also two user can have the same name that was why I put no restrictions on that too. If you want me to change it, I will do that once u give me the go ahead. 
> Thanks sir

Oh, I see what you mean in the transactions but there is a standard for that, that will be to paginate the response, but that's a bit complex so I didn't state that it should be implemented, so go ahead and return the actual transactions.

For the multi accounts same name, although rare there are cases where some people have the same name and that's why banks also collect a middle name. In this case I was flexible and some people did full name while some others did First and last name like you have, so since it's flexible should add the validation.

File: restapi.js (hunk 3 above)
Comment (joshuaunity) [reply restapi.js:37]: Hi @IGNORAMUS1 I tried this multiple times but it's still the same, looking at the code the only thing here is the unique field you added which will technically work but will cause the server to break, there is a better way to go about this by simply doing a DB lookup if the exact first and last name exists. You can go ahead and implement this yourself as I will be merging this.

**Pull request merged**